pkg/networkservice/nsmonitor/client_test.go

```diff
@@ -0,0 +1,339 @@
+)
+
+const (
+	testInode1 = "inode://4/4026534206"
+	testInode2 = "inode://4/4026534149"
+)
+
+type testMonitor struct {
+	inodes                     []string
+	watchShouldCloseMonitoing  bool
+	watchShouldCloseConnection bool
+}
+
+func (m *testMonitor) Watch(ctx context.Context, inodeURL string) <-chan struct{} {
+	m.inodes = append(m.inodes, inodeURL)
+
+	result := make(chan struct{}, 1)
+	if m.watchShouldCloseConnection {
+		result <- struct{}{}
+		close(result)
+	} else if m.watchShouldCloseMonitoing {
+		close(result)
+	}
+	return result
+}
+
+func Test_Client_DontFailWhenNoInode(t *testing.T) {
+	t.Cleanup(func() { goleak.VerifyNone(t) })
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+
+	monitor := testMonitor{}
+	supplyMonitor := func(c context.Context) nsmonitor.Monitor { return &monitor }
+
+	client := chain.NewNetworkServiceClient(
+		metadata.NewClient(),
+		begin.NewClient(),
+		nsmonitor.NewClient(ctx, nsmonitor.WithSupplyMonitor(supplyMonitor)),
+	)
+
+	// no inodeURL in parameters
+	_, err := client.Request(ctx, &networkservice.NetworkServiceRequest{
+		Connection: &networkservice.Connection{
+			Id: uuid.New().String(),
+			Mechanism: &networkservice.Mechanism{
+				Parameters: map[string]string{},
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.Empty(t, monitor.inodes)
+
+	// no mechanism parameters
+	_, err = client.Request(ctx, &networkservice.NetworkServiceRequest{
+		Connection: &networkservice.Connection{
+			Id:        uuid.New().String(),
+			Mechanism: &networkservice.Mechanism{},
+		},
+	})
+	require.NoError(t, err)
+	require.Empty(t, monitor.inodes)
+
+	// no mechanism
+	_, err = client.Request(ctx, &networkservice.NetworkServiceRequest{
+		Connection: &networkservice.Connection{
+			Id: uuid.New().String(),
+		},
+	})
+	require.NoError(t, err)
+	require.Empty(t, monitor.inodes)
+}
+
+func Test_Client_MonitorOnceSameConnectionId(t *testing.T) {
+	t.Cleanup(func() { goleak.VerifyNone(t) })
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+
+	monitor := testMonitor{}
+	supplyMonitor := func(c context.Context) nsmonitor.Monitor { return &monitor }
+
+	client := chain.NewNetworkServiceClient(
+		metadata.NewClient(),
+		begin.NewClient(),
+		nsmonitor.NewClient(ctx, nsmonitor.WithSupplyMonitor(supplyMonitor)),
+	)
+
+	request := &networkservice.NetworkServiceRequest{
+		Connection: &networkservice.Connection{
+			Id: uuid.New().String(),
+			Mechanism: &networkservice.Mechanism{
+				Parameters: map[string]string{
+					common.InodeURL: testInode1,
+				},
+			},
+		},
+	}
+
+	_, err := client.Request(ctx, request)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(monitor.inodes))
+	require.Equal(t, testInode1, monitor.inodes[0])
+
+	_, err = client.Request(ctx, request)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(monitor.inodes))
+	require.Equal(t, testInode1, monitor.inodes[0])
```

Comment:
The test can be unstable. Adding inodes for the test monitor is going in the goroutine. Please consider sync access. Moreover, add a check that monitor has handled watch operation to avoid race situations.